// @nx/vite/src/utils/versions.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    nxVersion: function() {
        return nxVersion;
    },
    viteVersion: function() {
        return viteVersion;
    },
    vitestVersion: function() {
        return vitestVersion;
    },
    vitestUiVersion: function() {
        return vitestUiVersion;
    },
    vitePluginReactVersion: function() {
        return vitePluginReactVersion;
    },
    vitePluginReactSwcVersion: function() {
        return vitePluginReactSwcVersion;
    },
    viteTsConfigPathsVersion: function() {
        return viteTsConfigPathsVersion;
    },
    jsdomVersion: function() {
        return jsdomVersion;
    },
    vitePluginDtsVersion: function() {
        return vitePluginDtsVersion;
    },
    happyDomVersion: function() {
        return happyDomVersion;
    },
    edgeRuntimeVmVersion: function() {
        return edgeRuntimeVmVersion;
    },
    vitestCoverageC8Version: function() {
        return vitestCoverageC8Version;
    },
    vitestCoverageIstanbulVersion: function() {
        return vitestCoverageIstanbulVersion;
    }
});
const nxVersion = require('../../package.json').version;
const viteVersion = '^4.3.9';
const vitestVersion = '^0.32.0';
const vitestUiVersion = '^0.32.0';
const vitePluginReactVersion = '^4.0.0';
const vitePluginReactSwcVersion = '^3.3.2';
const viteTsConfigPathsVersion = '^4.2.0';
const jsdomVersion = '~22.1.0';
const vitePluginDtsVersion = '~2.3.0';
const happyDomVersion = '~9.20.3';
const edgeRuntimeVmVersion = '~3.0.2';
const vitestCoverageC8Version = '^0.32.0';
const vitestCoverageIstanbulVersion = '^0.32.0';
